Cypress test for MM-T2368 'Fixed width center' setting does not cause scroll pop and breaks post display

furqanmlk opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
Added new test case MM-T2368 
This test verifies that Fixed width center settings does not cause scroll pop and breaks post display
<!--

-->

#### Ticket Link
https://automation-test-cases.vercel.app/test/MM-T2368
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.
-->

mattermod commented:
Thanks @furqanmlk for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

ali-farooq0 commented:
/check-cla

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,73 @@
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
```
Can you please remove the prod tag on the test. This is only added by the QA team once the test works fine continuously for 5 days.

furqanmlk replied:
:heavy_check_mark:

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,73 @@
+    });
+
+    it('MM-T2368 Fixed Width setting should not scroll pop and display posts properly', () => {
+        // Creating some post to verify scroll pop and Posts views
```
Suggested minor change:

`// # Creating some post to verify scroll pop and Posts views`

Note: Each step is prepended with #

furqanmlk replied:
:heavy_check_mark:

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,73 @@
+        cy.postMessage('This is the last post');
+
+        // # Switch the account settings for the test user to enable Fixed width center
+        cy.get('#headerInfo > .style--none').click();
+        cy.get('#accountSettings > .style--none > .MenuItem__primary-text').click();
+        cy.get('#accountSettingsModalLabel > span').should('contain', 'Account Settings');
```
There is a helper function which you can use:

```
        // # Go to Account Settings
        cy.toAccountSettingsModal();
```

furqanmlk replied:
:heavy_check_mark:

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,73 @@
+        cy.get('#accountSettings > .style--none > .MenuItem__primary-text').click();
+        cy.get('#accountSettingsModalLabel > span').should('contain', 'Account Settings');
+        cy.get('#displayButton', {timeout: 500000}).click();
+        cy.get('#channel_display_modeEdit > span').click();
+        cy.get('#settingTitle > span').should('contain', 'Channel Display');
```
Using the ID of the element should be good enough. You can ignore the `> span` and make it simpler.

furqanmlk replied:
:heavy_check_mark:

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,73 @@
+        cy.get('input#channel_display_modeFormatB').click();
+        cy.get('input#channel_display_modeFormatB').next().should('contain', 'Fixed width, centered');
+        cy.get('#saveSetting').click();
+        cy.get('#accountSettingsHeader > .close > [aria-hidden="true"]').click();
```
You can make it simpler by using the following locator:
`cy.get('#accountSettingsHeader > .close').click();`

furqanmlk replied:
:heavy_check_mark:

srkgupta reviewed (changes requested):
Thanks @furqanmlk. Overall the test looks good. However I have added some suggestions. 

Also, the test does not validates if the layout of the posts changes or not when fixed width is selected. This is how the page looks when the option "Fixed width, centered" is selected

![Screenshot 2020-09-16 at 7 13 01 PM](https://user-images.githubusercontent.com/1429138/93345705-bff81480-f850-11ea-9197-8958b7f6de36.png)

And this is how it looks when the option "Full width" is selected
![Screenshot 2020-09-16 at 7 13 12 PM](https://user-images.githubusercontent.com/1429138/93345742-cd150380-f850-11ea-8adc-85bde2f91217.png)

Hence can you please add some assertion to verify the changes. You can verify the new classes added to the elements when the "Fixed width, centered" option is selected.

furqanmlk commented:
> Thanks @furqanmlk. Overall the test looks good. However I have added some suggestions.
> 
> Also, the test does not validates if the layout of the posts changes or not when fixed width is selected. This is how the page looks when the option "Fixed width, centered" is selected
> 
> ![Screenshot 2020-09-16 at 7 13 01 PM](https://user-images.githubusercontent.com/1429138/93345705-bff81480-f850-11ea-9197-8958b7f6de36.png)
> 
> And this is how it looks when the option "Full width" is selected
> ![Screenshot 2020-09-16 at 7 13 12 PM](https://user-images.githubusercontent.com/1429138/93345742-cd150380-f850-11ea-8adc-85bde2f91217.png)
> 
> Hence can you please add some assertion to verify the changes. You can verify the new classes added to the elements when the "Fixed width, centered" option is selected.

Thank you so much Gupta for your excellent feedback.
I have added the assertions to verify changes by "Fixed width,centered" option .

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+    let testChannel;
+    let otherUser;
+
+    beforeEach(() => {
```
nit: Let's start with `before` until there's multiple test blocks where each require something before actual testing.

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+            testTeam = team;
+            testChannel = channel;
+
+            cy.apiCreateUser().then(({user: user2}) => {
+                otherUser = user2;
+
+                cy.apiAddUserToTeam(testTeam.id, otherUser.id).then(() => {
+                    cy.apiAddUserToChannel(testChannel.id, otherUser.id);
+                });
+            });
```
Seems like `otherUser` is not used elsewhere.

furqanmlk replied:
:heavy_check_mark: 
I was planning to have different post from other user, but I think we don't need it.

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+        });
+    });
+
+    it('MM-T2368 Fixed Width setting should not scroll pop and display posts properly', () => {
```
nit: Please change title to match the written test case.

```suggestion
    it('MM-T2368 Fixed width', () => {
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+    });
+
+    it('MM-T2368 Fixed Width setting should not scroll pop and display posts properly', () => {
+        // # Creating some post to verify scroll pop and Posts views
+        cy.postMessage('This is the first post');
+        Cypress._.times(18, (postIndex) => {
+            cy.postMessage(`p-${postIndex + 1}`);
+        });
+        cy.postMessage('This is the last post');
```
It's not specifically stated in the test case but like I mentioned in our DM, "scroll pop" is an internal term that refers to a sudden change of height from initial to final render that causes each post to change render position over time, and is noticeable when some posts have image, link preview, markdown in-line image and others.  With that, please add posting of message with:
- file attachment -- each post for `jpg-image-file.jpg`, `gif-image-file.gif`, `mp3-audio-file.mp3`. Note that these file names are already in the fixture folder.
- link preview using `https://www.bbc.com/news/uk-wales-45142614`, and
- markdown image using `![gif](http://i.giphy.com/xNrM4cGJ8u3ao.gif)`.

For these to open previews, you'll need to update config with `EnableLinkPreviews: true` (see other spec files for reference).
Also, please use taller view port, maybe 1300x2000, so many posts are visible within the window.

furqanmlk replied:
:heavy_check_mark: 
I have added image, link preview, markdown in-line image in posts
Verifying before and after height of these posts

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+
+        // # Switch the account settings for the test user to enable Fixed width center
+        cy.toAccountSettingsModal();
+        cy.get('#displayButton', {timeout: 500000}).click();
+        cy.get('#channel_display_modeEdit').click();
+        cy.get('#settingTitle').should('contain', 'Channel Display');
+        cy.get('input#channel_display_modeFormatB').click();
+        cy.get('input#channel_display_modeFormatB').next().should('contain', 'Fixed width, centered');
+        cy.get('#saveSetting').click();
+        cy.get('#accountSettingsHeader > .close').click();
```
- `500000` is huge timeout. Also you may use constants at `/fixtures/timeouts`.
- Since it shows account settings modal, queries to each element is better (scoped) within the modal -- should use `.within`.
- We're now using `testing-library` as much as possible. So, please use [api queries](https://testing-library.com/docs/dom-testing-library/api-queries) whenever possible. So that queries is represented on how the actual user sees is and not the underlying element IDs, CSS class, etc.

Something like:
```javascript
cy.get('#accountSettingsModal').should('be.visible').within(() => {
    cy.findByText('Display').click();
    cy.findByText('Channel Display').click();
    // do others ...
});
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+        cy.visit(`/${testTeam.name}/channels/${testChannel.name}`);
+
+        // * Verify there is no scroll pop
+        cy.get('button[aria-label="sysadmin"]').eq(0).should('be.visible');
```
May use `cy.findByLabelText`

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,74 @@
+            });
+        });
+
+        cy.get('[data-testid="postContent"].post__content.center').should('have.length', '21').each(($el) => {
```
May use May use `cy.findByTestId`

furqanmlk replied:
:heavy_check_mark:

saturninoabril reviewed (changes requested):
Thanks @furqanmlk. Please see comments.

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+        getThumbnailPost().eq(0).invoke('height').as('initialThumbnailHeight');
+        getAttachmentPost().eq(0).invoke('height').as('initialAttachmentHeight');
+        getInlinImgPost().eq(0).invoke('height').as('initialInlineImgHeight');
+        getFirstTextPost().eq(0).invoke('height').as('initial1stPostHeight');
```
nit: for consistency with "last"
```suggestion
        getFirstTextPost().eq(0).invoke('height').as('initialFirstPostHeight');
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+        getAttachmentPost().eq(0).invoke('height').as('initialAttachmentHeight');
+        getInlinImgPost().eq(0).invoke('height').as('initialInlineImgHeight');
+        getFirstTextPost().eq(0).invoke('height').as('initial1stPostHeight');
+        getLastTextPost().eq(0).invoke('height').as('initiallastPostHeight');
```
nit: for consistency using camelCase
```suggestion
        getLastTextPost().eq(0).invoke('height').as('initialLastPostHeight');
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+        return cy.get('.attachment__image');
+    };
+
+    const getInlinImgPost = () => {
```
Seems like typo?
```suggestion
    const getInlineImgPost = () => {
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+
+    const getUserNameTitle = () => {
+        return cy.findAllByLabelText('sysadmin');
+    };
+    const getThumbnailPost = () => {
```
```suggestion
    };

    const getThumbnailPost = () => {
```

furqanmlk replied:
:heavy_check_mark:

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+    });
+
+    const getUserNameTitle = () => {
+        return cy.findAllByLabelText('sysadmin');
```
Why not `cy.findByLabelText`?

furqanmlk replied:
cy.findByLabelText throwing error saying found multiple elements

![image](https://user-images.githubusercontent.com/6909890/93692602-1613c480-faaa-11ea-9fd5-dfb7508dbf2f.png)

saturninoabril commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,131 @@
+            cy.get('@initialThumbnailHeight').then((originalHeight) => {
+                getThumbnailPost().should('have.length', '2').and('have.css', 'height', originalHeight + 'px');
+            });
+            cy.get('@initialInlineImgHeight').then((originalHeight) => {
+                getInlinImgPost().should('have.css', 'height', (originalHeight + 2) + 'px');
+            });
+            cy.get('@initialAttachmentHeight').then((originalHeight) => {
+                getAttachmentPost().should('have.css', 'height', (originalHeight + 2) + 'px');
+            });
+
+            cy.get('@initialGifPostHeight').then((originalHeight) => {
+                getGifPost().should('have.css', 'height', originalHeight + 'px');
+            });
+
+            cy.get('@initialJpgPostHeight').then((originalHeight) => {
+                getJpgPost().should('have.css', 'height', originalHeight + 'px');
+            });
```
nit: spacing is not consistent here.

furqanmlk replied:
:heavy_check_mark:

saturninoabril reviewed (changes requested):
@furqanmlk Thanks for updating. However, still failing consistently on me with random errors (see screenshots). Could you confirm if test is passing on you.
Also, I don't see if the view port was changed.

<img width="627" alt="Screen Shot 2020-09-20 at 6 07 52 AM" src="https://user-images.githubusercontent.com/5334504/93690445-0f696980-fb0b-11ea-8160-722c3400cf86.png">
<img width="626" alt="Screen Shot 2020-09-20 at 6 09 31 AM" src="https://user-images.githubusercontent.com/5334504/93690448-12fcf080-fb0b-11ea-8e86-fe98dc726762.png">

furqanmlk commented:
I ran this test case 7 times with out any error
![image](https://user-images.githubusercontent.com/6909890/93693714-44000580-fab8-11ea-912f-76b5097576a1.png)


![image](https://user-images.githubusercontent.com/6909890/93693717-4cf0d700-fab8-11ea-9f44-bddaee72e8e2.png)

saturninoabril commented:
@furqanmlk Thanks for updating. Still failed on me but I'll let other reviewers to check as it might be only me.

srkgupta reviewed (changes requested):
@furqanmlk, the test is failing exactly at the same step as mentioned by @saturninoabril. I believe this is because as shown in this image, the word sysadmin is actually not visible because of the other posts and images. Can you please recheck this once again possibly by resizing your browser? 

![Screenshot 2020-09-24 at 6 28 44 PM](https://user-images.githubusercontent.com/1429138/94148305-150cd980-fe94-11ea-9a40-91271b07e75c.png)

furqanmlk commented:
@srkgupta thanks for double checking 

I found the real issue, it issue was not viewPort as I am not asserting any element visibility.
The issue was child elements of #post-list are taking time to appear.

srkgupta commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,160 @@
+    let testChannel;
+
+    before(() => {
+        cy.viewport(1800, 2000);
```
0/5 to set such a high view port unless its absolutely necessary. Currently the test works fine even without setting this view port. Hence can we remove this.

furqanmlk replied:
@srkgupta thanks, 
that's exactly my observation was.
I have reverted View-port

srkgupta reviewed (approved):
Thanks @furqanmlk. Tested and the test passed now. Just got one minor suggestion. The rest looks fine for me.

prapti commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,158 @@
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Group: @scroll
```
Suggesting an extra line after the `// Group` line:
```suggestion
// Group: @scroll

```

furqanmlk replied:
:heavy_check_mark:

prapti commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,158 @@
+        cy.postMessage(gifLink);
+        cy.getLastPostId().as('gifLinkPostId');
+        const commonTypeFiles = ['jpg-image-file.jpg', 'gif-image-file.gif', 'mp3-audio-file.mp3', 'mpeg-video-file.mpg'];
+        commonTypeFiles.forEach((file) => {
```
A one line comments would be great to make reading the test simple.

furqanmlk replied:
:heavy_check_mark:

prapti commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,158 @@
+        cy.postMessage('This is the last post');
+        cy.getLastPostId().as('lastPostId');
+
+        getUserNameTitle().eq(0).invoke('height').as('initialUserNameHeight');
```
Again, a one line comment to understand this block of action would be great.

furqanmlk replied:
:heavy_check_mark:

prapti commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,158 @@
+        return cy.findAllByLabelText('sysadmin');
+    };
+
+    const getMp3Post = () => {
```
Could you please add a comment to explain what this section is doing?

furqanmlk replied:
:heavy_check_mark:

prapti commented on e2e/cypress/integration/scroll/fixed_width_spec.js:
```diff
@@ -0,0 +1,158 @@
+        });
+    };
+
+    const getMpgPost = () => {
```
Requesting a comment again :)

furqanmlk replied:
:heavy_check_mark:

prapti reviewed (changes requested):
Hi @furqanmlk! Thank you so much for contributing this test, and working on all the suggestions made by Rohitesh and Saturn so far!
The test looks good in what it's doing, and is also passing 🎉 
One request I'd like to make before approving is to please add more comments. This will help us ensure that whoever is looking at this test file outside our group will find following it easy. I've left some inline suggestions where the comments would be great. Please add more such comments (at least one explanatory comment for a common section) so that it's easier for anyone to understand.
Thank you!

furqanmlk reviewed (commented):
@prapti thanks for the feedback
I always like to write the comments but I checked and realized  most of the test cases do not have comments other than #(for test step) or * (test assertion).
I have added comments as you suggested :)

saturninoabril reviewed (approved):
Thanks for updating @furqanmlk! Tested and passed. LGTM.

prapti reviewed (approved):
Thank you so much @furqanmlk! Looks great!
Approving.

prapti commented:
/update-branch

prapti commented:
/update-branch
